Block editor: Hotfix performance regression

mmtr opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/98377
See p1737631296756739-slack-C02FMH4G8

## Proposed changes:

Overrides the `wp-block-editor` script with a custom version that includes these changes (props to @youknowriad):

```diff
diff --git a/packages/block-editor/src/store/selectors.js b/packages/block-editor/src/store/selectors.js
index 22d725bbcd6..fed2331edc6 100644
--- a/packages/block-editor/src/store/selectors.js
+++ b/packages/block-editor/src/store/selectors.js
@@ -1626,15 +1626,22 @@ const isBlockVisibleInTheInserter = (
 		Array.isArray( blockType.parent ) ? blockType.parent : []
 	).concat( Array.isArray( blockType.ancestor ) ? blockType.ancestor : [] );
 	if ( parents.length > 0 ) {
-		const rootBlockName = getBlockName( state, rootClientId );
 		// This is an exception to the rule that says that all blocks are visible in the inserter.
 		// Blocks that require a given parent or ancestor are only visible if we're within that parent.
-		return (
-			parents.includes( 'core/post-content' ) ||
-			parents.includes( rootBlockName ) ||
-			getBlockParentsByBlockName( state, rootClientId, parents ).length >
-				0
-		);
+		if ( parents.includes( 'core/post-content' ) ) {
+			return true;
+		}
+		let current = rootClientId;
+		let hasParent = false;
+		do {
+			const currentName = getBlockName( state, current );
+			if ( parents.includes( currentName ) ) {
+				hasParent = true;
+				break;
+			}
+			current = state.blocks.parents.get( current );
+		} while ( current );
+		return hasParent;
 	}
 
 	return true;
```

This way, we can prevent a performance degradation that started with https://github.com/WordPress/gutenberg/pull/67734 which is currently present in Gutenberg v19.9.0, v20.0.0, and v20.1.0, and particularly on plugins that register blocks like Jetpack or WooCommerce.

This is especially relevant for Dotcom sites since the Gutenberg version has been pinned to v20.0.0 and there are no plans currently to update it (p58i-jaO-p2).

The custom `wp-block-editor` script has been built for each of the affected Gutenberg versions.

**Before**


https://github.com/user-attachments/assets/8e04fb13-3372-482b-9124-ed6f8071d147

**After**


https://github.com/user-attachments/assets/a3c2bf77-fcb9-4dd7-9499-2c860ebcd2fb




### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
- Apply these changes to an Atomic site
- Install WooCommerce and activate it
- Open the site editor
- Edit the navigation block and add some submenu items
- Save the changes and reload
- When hovering a menu, make sure the submenu appears immediately
- When selecting a menu from the list view, make sure the menu is selected immediately 
- Test with different Gutenberg versions and with Gutenberg inactive, and make sure everything works as expected

mmtr committed:
Block editor: Hotfix performance regression

mmtr committed:
Regenerate files

mmtr committed:
Fix path

jeherve commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -1404,3 +1412,37 @@ public static function register_block_metadata_collection() {
+		return;
+	}
+
+	if ( ! in_array( GUTENBERG_VERSION, array( '19.9.0', '20.0.0', '20.1.0' ), true ) ) {
```
Do we need to account for `GUTENBERG_DEVELOPMENT_MODE` as well?

mmtr replied:
Yeah, good idea. Otherwise Gutenberg developers would have unexpected side effects if they also have Jetpack active. Done in https://github.com/Automattic/jetpack/pull/41309/commits/7da9eaa37701e2f8b05f495585a2ed0be52aaae7

jeherve commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -1404,3 +1412,37 @@ public static function register_block_metadata_collection() {
+	gutenberg_override_script(
+		$scripts,
+		$handle,
+		Assets::get_file_url_for_environment(
+			'_inc/block-editor-hotfix-' . GUTENBERG_VERSION . '.min.js',
+			'_inc/block-editor-hotfix-' . GUTENBERG_VERSION . '.min.js'
```
Given that we only have minified files, would it be possible to simplify things and remove usage of the `get_file_url_for_environment` method?

mmtr replied:
Done in https://github.com/Automattic/jetpack/pull/41309/commits/b5d48b1ef5030a124fc7d04af7d79abac97e33cb

jeherve commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -1404,3 +1412,37 @@ public static function register_block_metadata_collection() {
+		return;
+	}
+
+	$path         = gutenberg_dir_path() . 'build/block-editor/index.min.js';
```
Phan isn't too happy about this, so you'll probably have to make an update to the config.

mmtr replied:
Added Gutenberg stubs in https://github.com/Automattic/jetpack/pull/41309/commits/f6c40cca183fee755ce88834ab70447a18131414

mmtr committed:
Bail when running a local dev build of Gutenberg

mmtr committed:
Simplify file path

mmtr committed:
Add Gutenberg stubs

mmtr committed:
Merge remote-tracking branch 'origin/trunk' into hotfix/block-editor-performance

mmtr committed:
Update Jetpack Phan config with Gutenberg stubs

mmtr committed:
Load Gutenberg stubs

jeherve reviewed (commented):
This tests well for me. I would have a couple of last remarks, but no blockers:

- What would you think about making the function name, `jetpack_hotfix_block_editor_script`, a bit more specific about the versions / the problem it targets, so we know what this was about in a few years. Maybe the docblock could include more information?
- Maybe that docblock could mention how the min. files were created? Or maybe just point to this PR for more information?
- The file is currently loaded even if Jetpack is not connected to WordPress.com yet, or if Jetpack's own block bundle is not loaded because the feature has been disabled. I think that's fine, the hotfix would benefit everyone anyway, but I wanted to make sure that was intentional.
- In #41333 we ship the hotfix only in mu-wpcom. If the performance impact touches everyone, regardless of the platform, I think it makes sense to ship it in Jetpack for now, but maybe there should be a clearer way for folks reviewing changes like this to understand why a hotfix goes in one plugin vs. the other.

youknowriad commented:
Looks like a fix for this is coming upstream, I wonder if it's worth keeping.

mmtr commented:
> Looks like a fix for this is coming upstream, I wonder if it's worth keeping.

We pinned Gutenberg v20 in Dotcom, so unless we revert that decision, I think we're going to need this hotfix. See p9oQ9f-2Bq-p2#comment-3374

mmtr committed:
Rename function and update docblock

mmtr committed:
Don't hotfix if blocks module is not active

mmtr commented:
> What would you think about making the function name, jetpack_hotfix_block_editor_script, a bit more specific about the versions / the problem it targets, so we know what this was about in a few years. Maybe the docblock could include more information?
> Maybe that docblock could mention how the min. files were created? Or maybe just point to this PR for more information?

Good callout. Renamed the function and updated the docblock in https://github.com/Automattic/jetpack/pull/41309/commits/b02606a19b3c419e1989cd43fff06a0583d2781d

> The file is currently loaded even if Jetpack is not connected to WordPress.com yet, or if Jetpack's own block bundle is not loaded because the feature has been disabled. I think that's fine, the hotfix would benefit everyone anyway, but I wanted to make sure that was intentional.

Definitely, not intentional. Although the issue is indeed affecting everyone, is almost unnoticeable if the site doesn't have a plugin registering additional blocks. Since Jetpack is one of these plugins, I think it'd be better to target this hotfix to only when the Jetpack blocks are available.

In https://github.com/Automattic/jetpack/pull/41309/commits/e22639ac613d8222946b082a1aa89817a8462094, I included an additional check for that.

> In #41333 we ship the hotfix only in mu-wpcom. If the performance impact touches everyone, regardless of the platform, I think it makes sense to ship it in Jetpack for now, but maybe there should be a clearer way for folks reviewing changes like this to understand why a hotfix goes in one plugin vs. the other.

Yup, I'm aware of this inconsistency. I don't know why #41333 only hotfixed Dotcom sites, maybe the issue was specific to Dotcom sites. If that's not the case, I think we should apply to Jetpack sites too (cc @fushar). I briefly mentioned this too in p9oQ9f-2Bq-p2#comment-3374, since we're going to need more alignment with other folks (e.g. WooCommerce).

mmtr committed:
Fix version in docblock

fushar commented:
> I don't know why https://github.com/Automattic/jetpack/pull/41333 only hotfixed Dotcom sites, maybe the issue was specific to Dotcom sites. If that's not the case, I think we should apply to Jetpack sites too (cc @fushar).

My understanding is that we need to hotfix to Dotcom (only) because the GB version is pinned there. Other hosts are free to upgrade the GB plugin whenever a fix is available upstream. Am I missing something?

mmtr commented:
> My understanding is that we need to hotfix to Dotcom (only) because the GB version is pinned there. Other hosts are free to upgrade the GB plugin whenever a fix is available upstream. Am I missing something?

That makes sense, yeah. Self-hosted sites are also more unlikely to have the Gutenberg plugin active. I guess I lost perspective because I started to work on this before the issue was fixed upstream, so I was considering a worst-case scenario where the issue was solved in the short term.

I'll move the changes here to `jetpack-mu-wpcom`.

anomiex commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -73,9 +73,8 @@ class Jetpack_Gutenberg {
-	 * @since 8.3.0
-	 *
 	 * @return boolean True if the version of gutenberg required by the block or plugin is available.
+	 * @since 8.3.0
```
Why all the moving around of stuff like this? All these extraneous changes make this PR hard to review, and a few like this one seem contrary to how we normally do things.

mmtr replied:
This was done automatically by the linter when I committed the files, I didn't change this.

anomiex replied:
Was it the linter, or your IDE? When I run the linting locally on trunk it doesn't make these changes, and if I check out this PR and revert them it doesn't redo them when committing. Even when `git reset --soft` back to the merge-base with trunk and recommitting. 🤷

mmtr replied:
Ha, you are spot on – it was my IDE 🤦

anomiex commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -262,7 +262,7 @@ public static function preset_exists( $preset ) {
 	public static function get_preset( $preset ) {
 		return json_decode(
-			// phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
```
Particularly this, why the bad indent?

mmtr replied:
No idea, I guess the linter didn't like this and change it.

anomiex reviewed (commented):
> I'll move the changes here to jetpack-mu-wpcom.

Glad I looked at the comments before suggesting exactly that! :+1: 😀

Even better would be to bump Gutenberg to a fixed version before freezing it, but 🤷.

mmtr committed:
Merge remote-tracking branch 'origin/trunk' into hotfix/block-editor-performance

mmtr committed:
Revert changes in the Jetpack plugin

mmtr committed:
Remove Jetpack changelog

mmtr committed:
Revert Phan changes for the Jetpack plugin

mmtr committed:
Move hotfix to jetpack-mu-wpcom

mmtr committed:
Revert automated linter changes

mmtr committed:
Fix Phan errors

mmtr committed:
Remove Phan suppression

anomiex reviewed (commented):
Looks reasonable to me, assuming we won't update Gutenberg directly to v20.2.0 on wpcom.

Haven't tested it myself though.

dsas reviewed (approved):
Tests fine on atomic for me, I can't repro the original issue with child blocks either. I can't repro the slowness on simple.

Perhaps we should commit the original source file here as well as the minified ones but given the hotfix is already merged upstream I guess it's unlikely to be useful.

youknowriad commented:
I've been thinking about this change and I'm a bit concerned about it. the main issue is that people are going to use Jetpack with future Gutenberg versions that might introduce new functions and APIs to the block-editor package. When that happens, Jetpack will replace block-editor with an old patched version which means potential breakage. So unless we keep patching this package with every Gutenberg release, it's going to be very hard to keep this fix in place.

Given that the fix landed in Gutenberg upstream, I think we should just revert this PR.

anomiex commented:
> the main issue is that people are going to use Jetpack with future Gutenberg versions 

In the end, the fix isn't in Jetpack at all. It's in the mu-wpcom package that's only intended for use in Simple and Atomic.

> When that happens, Jetpack will replace block-editor with an old patched version which means potential breakage. 

I note the fix here checks for specific Gutenberg versions. Later versions won't be replaced.
https://github.com/Automattic/jetpack/blob/ba9862be0cd52801f10aba5f4b2245c040a012b5/projects/packages/jetpack-mu-wpcom/src/features/wpcom-hotfixes/wpcom-hotfixes.php#L69-L71

youknowriad commented:
> I note the fix here checks for specific Gutenberg versions. Later versions won't be replaced.

I had missed that, that eases my concerns a bit. Thank you.
